python/KAPEL.py

```diff
@@ -52,7 +52,8 @@ def __init__(self, queryRange, namespace):
         # (which takes a range and returns a scalar), and as a result get the whole metric set. Finally, use group_left for many-to-one matching.
         # https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators
         # https://prometheus.io/docs/prometheus/latest/querying/operators/#many-to-one-and-one-to-many-vector-matches
-        self.cputime = f'(max_over_time(kube_pod_completion_time{{namespace="{namespace}"}}[{queryRange}]) - max_over_time(kube_pod_start_time{{namespace="{namespace}"}}[{queryRange}])) * on (pod) group_left() max without (instance, node) (max_over_time(kube_pod_container_resource_requests{{resource="cpu", node != "", namespace="{namespace}"}}[{queryRange}]))'
+        self.cputime = f'max_over_time(container_cpu_usage_seconds_total{{namespace="{namespace}"}}[{queryRange}])'
```

Comment:
@mwestphall  This looks like  CPU usage time which is very different from the current cputime (CPU walltime) query which should remain unchanged for APEL accounting to keep working. Having CPU usage would be useful though.

Based on some testing this query will double-count all records because it looks like you get container-level results along with the top-level cgroup (presumably?) corresponding to the pod. The only way I can see to only get the right top-level cgroup is to select records without the `name` field. See attached image as an example.

Reply:
![Screenshot from 2024-05-24 18-20-25](https://github.com/rptaylor/kapel/assets/1686627/ae7f7af3-d6ff-4a4a-aaf5-33810bd5537f)

Reply:
Hmm also these metrics are being collected from a different source, it must be the cAdvisor metrics built into the kubelet, instead of the kube-state-metrics pod. I wonder if you might run into different kinds of issues similar to those avoided by  using `without (instance, node)` as explained in the comments.

Reply:
FYI this reports CPU usage for pods that are currently still running. A subsequent run would then report more usage for the same pod later.

Reply:
@rptaylor Good catch with the double counting of records, I didn't catch that. It looks like kubernetes creates a separate ["pause" container](https://kubernetes.io/docs/concepts/windows/intro/#pause-container) within each pod to hold the cgroup in case the other containers crash, that's likely what's leading to the duplicate records here. With regards to deduplication, do you think it would work to do a `by (pod)` rather than `without (<list of attributes>)`? In our prometheus instance, the record for the pause container differs from the record for the main container by several attributes (`without(name, id, image, namespace, container)` appears to deduplicate successfully)
![image](https://github.com/rptaylor/kapel/assets/15094432/46ecc653-32a9-4518-8aee-231d9e569cbb)

Reply:
I think `by pod` will hide the double counting without fixing it: https://stackoverflow.com/questions/51064821/difference-between-promql-by-and-without-unclear
Using by or without are different ways of aggregating results together, not deduplicating per se.
AFAICT based on the results I saw, the only way to be sure to count only the record corresponding to the parent cgroup (i.e. the overall pod itself without also including the component containers) would be to select items without a name field (maybe `name == ""` or `name=Null` would work? Not sure of the syntax), or parse the id field to make sure it matches the form of the parent cgroup of a pod (though yours looks different from mine, maybe it is a cgroups v1 vs v2 difference).

Another option could be to select only records that have a defined image field (i.e. all cgroups *except* the parent pod). Using without or by would make sense here to aggregate all those records, one for each real container, into one result..

Reply:
Just for clarification, the record that we want to report on is (in the example screenshot you sent) the second record with no name attribute? In that scenario, would attempting to use `max by (pod)` rather than `sum by (pod)` work, since it'd only take the highest CPU time from the nameless record rather than the lesser values from the other two? 

To further complicate things, which versions of kubernetes is your cluster using? In Tiger, prometheus is not reporting that nameless record, just one for the pause container and one for the pod's main container. Tiger is on version kubernetes 1.27, where it looks like `container_cpu_usage_seconds_total` is still an alpha feature. It doesn't become stable until 1.29. 
![image](https://github.com/rptaylor/kapel/assets/15094432/8f99c608-4752-49ff-a297-70acf599ebc7)

Reply:
Currently our clusters are v1.24 and as I understand cgroup v2 is promoted to stable in v1.25, so yes that could explain the difference.

Yes I think what you want is the CPU usage of the whole pod (all containers), whereas by its name `container_cpu_usage_seconds_total` is a container-based metric.
My interpretation of the screenshot I posted is: 13.14 for the main job container in the pod, .03 for the pause container, 13.16 total for the whole pod.   Some pods could have multiple workload containers.

My understanding of `max` is it selects the larger of two different metrics but in this case the only metric you're interested in is container_cpu_usage_seconds_total. I think the issue here is one metric, with multiple time series data points, distinguished by different labels. 

Hmm so maybe in cgroups v2, if the pod itself is not (mis)reported along with the containers in the container_cpu_usage_seconds_total metric, there should not be an issue of double counting. If so, based on your latest screenshot, if for every pod you only get the child containers, then adding them all together would be correct.

Reply:
Still looking into this, but the query that I ran a couple hours ago no longer returns any results. I believe there may be an additional issue that the metrics derived from cAdvisor aren't persisted in tiger for as long as those from kube-state-metrics, which will definitely cause additional issues.

Reply:
That query is now returning results again. I will definitely need to investigate the cause of that data absence though. It looks like you can aggregate `by` a single metric to get the largest of any number of records, grouped by the `by` parameter:
```
max_over_time(container_cpu_usage_seconds_total{namespace="uwdf-dev", pod="kapel-test-ep-rs62w"}[1d])

{container="osgvo-docker-pilot", cpu="total", endpoint="https-metrics", image="hub.opensciencegrid.org/opensciencegrid/osgvo-docker-pilot:23-release", namespace="uwdf-dev", node="tiger0011.chtc.wisc.edu", pod="kapel-test-ep-rs62w", service="kubelet", ...}
	49.119534462
{cpu="total", endpoint="https-metrics", image="k8s.gcr.io/pause:3.2", namespace="uwdf-dev", node="tiger0011.chtc.wisc.edu", pod="kapel-test-ep-rs62w", service="kubelet", ...}
	0.037396571
```

vs
```
max by (pod) max_over_time(container_cpu_usage_seconds_total{namespace="uwdf-dev", pod="kapel-test-ep-rs62w"}[1d])

{pod="kapel-test-ep-rs62w"}
	49.119534462
```
The `max by` takes the largest of any `container_cpu_usage_seconds_total` records where `pod = "kapel-test-ep-rs62w"`. This should give an accurate value in your cluster version since the whole-pod metric is always going to be the largest, but will slightly undercount in version 1.27 since it's excluding the pause container's small amount of cpu usage.

Reply:
I also noticed some irregularities regarding the time periods for which container_cpu_usage_seconds_total metrics for a given job were recorded, but couldn't narrow it down. Also if Prometheus is ever unavailable for a period of time you get a gap in the records.

In principle, some pods could have multiple worker containers and it would be good to account all of them, unless it is documented as a limitation of the Gratia CPU usage reporting in Kuantifier.

In any case the original point of this thread remains, this is a very different query than the CPU time needed by APEL. Can you make a new separate variable for this query?  Having both CPU wall time and CPU usage will be useful.

Reply:
@rptaylor oops, I made that change last week but never pushed it. The original metric should now be restored with a separate metric for the container_cpu_usage_seconds_total metric.

Reply:
From what I understood previously https://github.com/rptaylor/kapel/pull/54#discussion_r1619323129, I thought with cgroups v2, it turned out there would not be an issue of double counting after all? (The reason I thought there would is because of the cgroups v1 results I was seeing). You get two results, one for the pause container and one for the work container, and if you aggregate them together you get the right amount. I think it should be like this:
`sum by (pod) (max_over_time(container_cpu_usage_seconds_total{namespace="harvester",pod="grid-job-5143931-glsmj"}[1d]))`

If you do `max by (pod)` it means that only the largest CPU user of all containers in a pod will be accounted, so there would need to be a documented limitation that the results are not correct if a pod has multiple worker containers. This would make the accounting system unusable for any new job workflows that can take advantage of e.g. sidecar patterns in k8s, which are not possible in traditional batch systems.

In other words, if there is going to be a known limitation, I would rather it be that you get double counting if and only if the old cgroups v1 is used (i.e. using sum by pod), rather than only counting one container in a pod which uses the most CPU (i.e. using max by pod) in the modern cgroups v2. Does that make sense?

Reply:
@rptaylor yeah, that sounds good. `sum by (pod)` should be more accurate for v2 cgroups, `max by (pod)` was just an attempt to compromise between the v1 and v2 reporting. I'll update to use `sum by (pod)`.

Reply:
Nice. I noticed one other improvement, since the CPU usage should be a monotonically increasing cumulative value, you should be able to use `last_over_time` instead of max_over_time which should be much easier performance-wise on your Promethei, especially for long jobs. Based on my testing I confirmed the results look the same and for large time periods it was much faster.

https://prometheus.io/docs/prometheus/latest/querying/functions/#aggregation_over_time